src/python/pants/core/util_rules/external_tool.py

```diff
@@ -140,15 +133,13 @@ def generate_url(self, plat: Platform) -> str:
         os and arch default to those of the current system.
 
         Implementations should raise ExternalToolError if they cannot resolve the arguments
-        to a URL.  The raised exception need not have a message - a sensible one will be generated.
+        to a URL. The raised exception need not have a message - a sensible one will be generated.
         """
 
+    @abstractmethod
     def generate_exe(self, plat: Platform) -> str:
-        """Returns the archive path to the given version of the tool.
-
-        If the tool is downloaded directly, not in an archive, this can be left unimplemented.
-        """
```

Comment:
While this is less convenient that you must always specify the exe now, I think that it's good to be more explicit. Imo, it makes it more clear how `ExternalTool` works and makes it feel less magical.

Reply:
So, `bin/protoc` is still a relative path. The `./` is related to the executing environment in a way that I still don't fully understand, but it doesn't feel like something that should be included on relative paths in general. For example: it should be perfectly fine to relocate the executable before running it. And if you did that, you'd have to then re-add the `.` in order to execute it.

So it feels like this is something that we should do before executing the process iff the path is relative, or if we know that we're executing something inside the sandbox. Is there a clear place to do that here?

Reply:
> So it feels like this is something that we should do before executing the process iff the path is relative, or if we know that we're executing something inside the sandbox. Is there a clear place to do that here?

I don't think I understand what this means. Like, the call sites where the `Process` is set up should do this? Or the engine, perhaps?

Reply:
Call sites where the `Process` is set up, yes. I don't think it is possible to do it inside the engine, because it won't know whether you wanted a file from the PATH or a relative file. It's a callsite specific thing, essentially.